Allow creation of CustomActions

mikkeljohnsen opened the pull request.

mikkeljohnsen committed:
Allow creation of CustomActions

mikkeljohnsen commented:
```
    "custom_actions": [
    	{
    		"id": "ActionIdentifier",
    		"property": "",
    		"exe_command": "\"[INSTALLDIR]bin\\script.exe\"",
    		"execute": "deferred",
    		"return": "ignore",
    		"impersonate": "no",
    		"after": "InstallFiles"
    	}
    ],
```

mikkeljohnsen committed:
If more than one script, only have one 'InstallExecuteSequence' node

mikkeljohnsen committed:
Allow to also use 'Before'

jpakkane commented:
Looks good, but every new feature should come with a test project using it so we can ensure it works and does not break accidentally in the future. Please add one.

mikkeljohnsen commented:
Im going on  vaccation for 2 weeks. I will look at it after.

mikkeljohnsen committed:
Test for Custom Actions. When installing Windows reports error. Maybe it works with WiX on Windows

jpakkane commented on createmsi.py:
```diff
@@ -261,6 +267,21 @@ def create_registry_entries(self, comp, reg):
+            'Return': action['return'],
+            'Impersonate': action['impersonate'],
+        })
+        key = 'After' if action.has_key('after') else 'Before'
```
This fails to run. Dicts do not have a `has_key` method. The correct form is `if 'after' in action`.

mikkeljohnsen replied:
I was using python 2.7. I have changed to use the new way.

mikkeljohnsen committed:
Fix deprecated use of has_key. Add to run_tests

jpakkane commented:
The test fails with this with WiX:

```
C:\Users\IEUser\sharpkit\customactions\customaction.wxs(29) : error CNDL0006 :
The CustomAction/@Property attribute's value cannot be an empty string. If a value
is not required, simply remove the entire attribute.
```

You might also want to rebase this against master, since it now conflicts with the other PR that was merged. Thanks.

mikkeljohnsen committed:
Rebase with master
Create Property as cmd. This for now forces all exe command to use cmd.
Fix build with 'wixl' to use the arch.

mikkeljohnsen committed:
Use x86 when arch is 32

mikkeljohnsen committed:
Only create Property one time

mikkeljohnsen committed:
Property is hard coded, so do not use in example

mikkeljohnsen committed:
Remove Existing Products after InstallFinalize when using Custom Actions

jpakkane commented:
It still fails:

```
C:\Users\IEUser\sharpkit\customactions\customactions.wxs(5) : error LGHT0091 :
Duplicate symbol 'WixAction:InstallExecuteSequence/RemoveExistingProducts' found.
This typically means that an Id is duplicated. Check to make sure all your identifiers
of a given type (File, Component, Feature) are unique.
C:\Users\IEUser\sharpkit\customactions\customactions.wxs(27) : error LGHT0092 :
Location of symbol related to previous error.
```

mikkeljohnsen commented:
I don't see why it should be Duplicate. Here is what "wixl" generates:

Maybe Windows automatically creates a "RemoveExistingProducts" ?

```
<?xml version="1.0" ?>
<Wix xmlns="http://schemas.microsoft.com/wix/2006/wi">
 <Product Codepage="1252" Id="*" Language="1033" Manufacturer="Somebody" Name="Custom Action test" UpgradeCode="2DE3A7D5-0B2A-6A45-A93E-7700766A811F" Version="1.0.0">
  <Package Comments="Custom Action creation for executing script doing install" Compressed="yes" Description="Custom Action test 1.0.0 installer" Id="*" InstallerVersion="500" Keywords="Installer" Languages="1033" Manufacturer="Somebody" Platform="x64" SummaryCodepage="1252"/>
  <MajorUpgrade DowngradeErrorMessage="A newer version of Custom Action test is already installed."/>
  <Media Cabinet="customactions.cab" EmbedCab="yes" Id="1"/>
  <Directory Id="TARGETDIR" Name="SourceDir">
   <Directory Id="ProgramFiles64Folder">
    <Directory Id="INSTALLDIR" Name="customactions">
     <Component Guid="3462AB0D-42E1-444B-A765-453F53343CF9" Id="ApplicationFiles0" Win64="yes">
      <File Id="main_test.txt" Name="test.txt" Source="main/test.txt"/>
      <File Id="main_prog.bat" Name="prog.bat" Source="main/prog.bat"/>
     </Component>
    </Directory>
   </Directory>
   <Directory Id="ProgramMenuFolder"/>
  </Directory>
  <Property Id="WIXUI_INSTALLDIR" Value="INSTALLDIR"/>
  <Feature ConfigurableDirectory="INSTALLDIR" Description="The complete package" Display="expand" Id="Complete" Level="1" Title="Custom Action test 1.0.0">
   <Feature Absent="disallow" Description="This does nothing" Id="MainProgram" Level="1" Title="The Program">
    <ComponentRef Id="ApplicationFiles0"/>
   </Feature>
  </Feature>
  <InstallExecuteSequence>
   <RemoveExistingProducts After="InstallFinalize"/>
   <Custom Action="ActionIdentifier" After="InstallFiles"/>
  </InstallExecuteSequence>
  <Property Id="cmd" Value="cmd.exe"/>
  <CustomAction ExeCommand="/c &quot;[INSTALLDIR]prog.bat&quot;" Execute="deferred" Id="ActionIdentifier" Impersonate="no" Property="cmd" Return="ignore"/>
 </Product>
</Wix>
```

jpakkane commented:
> Maybe Windows automatically creates a "RemoveExistingProducts" ?

Maybe. But this still needs to be fixed before it can be merged.

Rakielle commented:
@jpakkane I want to try to fix the failing test, could you please give me any pointers?

jpakkane commented:
First you should clone the repo with the MR to your own machine, run the test suite and then verify that you get the same error.

Then it is a matter of examining the XML file that MsiCreator produces and trying to determine a) what the duplicate token is anc b) why it is getting generated. The error message's hint of "This typically means that an Id is duplicated. Check to make sure all your identifiers of a given type (File, Component, Feature) are unique." might be useful or it might not (I don't really know, this is the joy of debugging).

Rakielle commented:
@jpakkane Please, how do I run the test suite and is there a guide I can follow on how to use the project?

jpakkane commented:
There's no guide as such, but roughly:

- on Windows install Python and Visual Studio
- start the x64 native tools command prompt (from Visual Studio)
- `cd` into the directory where you have the checkout
- run `python run_tests.py`

kingmigo commented:
is it still open to fix the failed testing?

jpakkane commented:
Sure, this will be closed when the issue is fixed.

Neustradamus commented:
It has been solved with https://github.com/jpakkane/msicreator/pull/15?
